src/backend/drm/gbm/surface.rs

```diff
@@ -1,97 +1,146 @@
-use super::super::{Device, RawDevice, RawSurface, Surface};
+use super::super::{RawSurface, Surface};
 use super::Error;
 
-use drm::control::{connector, crtc, framebuffer, Device as ControlDevice, Mode};
+use drm::control::{connector, crtc, framebuffer, Mode};
 use failure::ResultExt;
-use gbm::{self, BufferObject, BufferObjectFlags, Format as GbmFormat, SurfaceBufferHandle};
+use gbm::{self, BufferObject, BufferObjectFlags, Format as GbmFormat};
 use image::{ImageBuffer, Rgba};
 
-use std::cell::{Cell, RefCell};
-use std::rc::Rc;
+use std::sync::{Arc, Mutex, atomic::{AtomicBool, Ordering}};
 
 use crate::backend::graphics::CursorBackend;
 
-pub(super) struct GbmSurfaceInternal<D: RawDevice + 'static> {
-    pub(super) dev: Rc<RefCell<gbm::Device<D>>>,
-    pub(super) surface: RefCell<gbm::Surface<framebuffer::Handle>>,
-    pub(super) crtc: <D as Device>::Surface,
-    pub(super) cursor: Cell<(BufferObject<()>, (u32, u32))>,
-    pub(super) current_frame_buffer: Cell<Option<framebuffer::Handle>>,
-    pub(super) front_buffer: Cell<Option<SurfaceBufferHandle<framebuffer::Handle>>>,
-    pub(super) next_buffer: Cell<Option<SurfaceBufferHandle<framebuffer::Handle>>>,
-    pub(super) recreated: Cell<bool>,
+pub struct Buffers {
+    pub(super) current_frame_buffer: Option<framebuffer::Handle>,
+    pub(super) front_buffer: Option<BufferObject<framebuffer::Handle>>,
+    pub(super) next_buffer: Option<BufferObject<framebuffer::Handle>>,
+}
+
+pub(super) struct GbmSurfaceInternal<S: RawSurface + 'static> {
```

Comment:
Because we do not share the underlying device with the surface anymore (see the changes to gbm/mod.rs), we do not need the Device as a generic anymore, but only the Surface. This makes the types easier to understand imo, so I took the chance to refactor this as well. (Same applies to `EGLStreamSurface`).